src/pdn/src/PdnGen.tcl

```diff
@@ -131,13 +131,14 @@ sta::define_cmd_args "define_pdn_grid" {[-name <name>] \
                                         [-halo <list_of_halo_values>] \
                                         [-pin_direction (horizontal|vertical)] \
                                         [-pins <list_of_pin_layers>] \
-                                        [-starts_with (POWER|GROUND)]}
+                                        [-starts_with (POWER|GROUND)] \
+                                        [-switch_cell <switch_cell_name>]}
 
 proc define_pdn_grid {args} {
   pdngen::check_design_state
 
   sta::parse_key_args "define_pdn_grid" args \
-    keys {-name -voltage_domains -orient -instances -cells -halo -pin_direction -pins -starts_with} \
+    keys {-name -voltage_domains -orient -instances -cells -halo -pin_direction -pins -starts_with -switch_cell} \
```

Comment:
Along with the cell, you will need a way to define the on/off conditions and pins on the celll.

Reply:
This PR is focussed on inserion of power switches into the design.

Beyond this, I agree there is a requirement to be able to define the structure of the power control network. This overlaps with some of the work being done by University of Minnesota. I'd be interested in your views on how the power control signal should be configured. Simple configurations like a star/high fanout net and daisy-chain should be easy to implement, but have their drawbacks (inrush too high and turn on too slow respectively). My approach for this is to have a few cells spread across the design that are configured to bring the rail up to 90% of its full value - only 5%-10% of the total number of power switches need to be turned on the avchieve this. Once the rail is at this value, the remaining power switches can be turned on all at the same time, without generating an inrush current spike. To my mind this "hybrid" scheme is the optimal approach.

The ability to define on/off conditions for power switches/power domains does not impact the placement of power switches in the design, but will impact verification of power domain boundaries, so I see it as outside the scope of this PR, and outside the scope of pdn.

This kind of on/off condition specification I would expect to be done with UPF commands, as this is a standard that defines these kind of interactions. The UPF specification defines a logical power switch for a power domain, but doesnt go into the detail of distributed coarse grain power switches. The closest it gets is the map_power_switch command, which in my view is justa way for UPF to reason about power switches in the design, rather than a way to drive the insertion of power switches in the first place.

Reply:
>  Simple configurations like a star/high fanout net and daisy-chain should be easy to implement, but have their drawbacks (inrush too high and turn on too slow respectively)

I think this would be dominated by the cells available in a given PDK. 

> The ability to define on/off conditions for power switches/power domains does not impact the placement of power switches in the design, but will impact verification of power domain boundaries, so I see it as outside the scope of this PR, and outside the scope of pdn.

I mildly disagree with, if the PR was only concerned with power switches already in the design (ie. they already have all appropriate nets attached), than I think it's correct. But because this is inserting new cells, those will need to get configured correctly and PDN is the correct place for that. One possible hack would be to "copy" the pin connectivity from one instance in that region to the next, but I would empathize the hack part there. If this PR is not ready to properly handle new cell insertion, then maybe that should be removed and handled separately?